compilers/concrete-compiler/compiler/include/concretelang/Analysis/TypeInferenceAnalysis.h

```diff
@@ -0,0 +1,940 @@
+class TypeInferenceAnalysisBase : public AnalysisT {
+public:
+  // Constructor passing on `constructorArgs` to the analysis class
+  template <typename... ConstructorArgTs>
+  TypeInferenceAnalysisBase(TypeResolver &resolver,
+                            mlir::DataFlowSolver &solver,
+                            ConstructorArgTs &&...constructorArgs)
+      : AnalysisT(solver, std::forward<ConstructorArgTs>(constructorArgs)...),
+        resolver(resolver) {}
+
+protected:
+  // Returns the types inferred for all values related to the
+  // operation `op` from the respective lattice values. For values,
+  // for which no type has been inferred so far, a new lattice with an
+  // empty inferred types is created.
+  LocalInferenceState getCurrentInferredTypes(mlir::Operation *op) {
+    LocalInferenceState map =
+        TypeInferenceUtils::getLocalInferenceState(op, [&](mlir::Value v) {
+          return this->template getOrCreate<TypeInferenceLattice>(v);
+        });
+
+    return map;
+  }
+
+  // Dumps the inferred type for all values related to `op` on
+  // `llvm::dbgs()`.
+  void debugPrintOp(mlir::Operation *op) {
+    LocalInferenceState inferredTypes = getCurrentInferredTypes(op);
+
+    llvm::dbgs() << "Inference state for " << op->getName() << " \n";
+    for (mlir::Value operand : op->getOperands()) {
+      InferredType operandType = inferredTypes.find(operand);
+      llvm::dbgs() << "  Operand: " << operandType << "\n";
+    }
+
+    for (mlir::Value result : op->getResults()) {
+      InferredType resultType = inferredTypes.find(result);
+      llvm::dbgs() << "  Result: " << resultType << "\n";
+    }
+
+    for (mlir::Region &r : op->getRegions()) {
+      for (mlir::Value v : r.getArguments()) {
+        InferredType argType = inferredTypes.find(v);
+        llvm::dbgs() << "  RegionArg: " << argType << "\n";
+      }
+    }
+  }
+
+  // Updates the lattice for the inferred type for the value `v` using
+  // its inferred type specified by `state`.
+  void update(LocalInferenceState &state, mlir::Value v) {
+    TypeInferenceLattice *lattice =
+        this->template getOrCreate<TypeInferenceLattice>(v);
+    auto latticeValue = lattice->getValue();
+    latticeValue.setType(state.find(v));
+    mlir::ChangeResult res = lattice->join(latticeValue);
+    this->propagateIfChanged(lattice, res);
+  }
+
+  // Updates the lattices for the inferred types for all values from
+  // `state` using the inferred types contained in `state`.
+  void update(LocalInferenceState &state, mlir::ValueRange values) {
+    for (mlir::Value v : values) {
+      update(state, v);
+    }
+  }
+
+  // Visits a single operation: looks up the lattice values for the
+  // values related to the operation, extracts the inferred types,
+  // invokes the type resolver and updates the lattice values with the
+  // newly inferred types.
+  void doVisitOperation(Operation *op) {
+    // Skip operations that do not use unresolved types
```

Comment:
Yes. For any operation that does not have an unresolved type, there is no need to propagate operand and result types, since they are already specified explicitly directly in the IR.